### PR TITLE
Start controllers from outside of Webots

### PR DESCRIPTION
**Pull request by omichel**
This PR aims at allowing users to start controller programs from outside of Webots.
That will help in the following use case:

- running a Webots controller from an external IDE or from a terminal (e.g., XCode, Eclipse, PyCharm, atom, python command line, etc.)
- debugging a controller by running it with debug tools (gdb, etc.).
- facilitate the integration with ROS and/or Cloud integration.

Main documentation: https://cyberbotics.com/doc/guide/running-extern-robot-controllers?version=feature-start-controllers-outside-of-webots

### CODE REVIEW

**Comment by omichel** [comment]
test me

**Comment by omichel** [comment]
test me

**Comment by omichel** [comment]
test me

**Comment by omichel** [comment]
test me

**Comment by omichel** [comment]
test me

**Comment by omichel** [comment]
test me

**Comment by omichel** [comment]
test me

**Comment by omichel** [comment]
test me

**Comment by omichel** [comment]
test me

**Comment by omichel** [comment]
👍

**Comment by omichel** [comment]
test me

**Comment by omichel** [comment]
test me

**Comment by omichel** [comment]
👍

**Comment by DavidMansolino** [comment]
:eyes:

File: change_logs/ChangeLog.html [hunk 1]
```diff
@@ -20,10 +20,14 @@ <h2><a href='https://cyberbotics.com/doc/blog/Webots-2018-b-release'>Webots R201
...
 <ul>
 <li>Bloom and Ambient Occlusion are now also available as effects for color cameras.</li>
+<<<<<<< HEAD
+<li>Extern robot controllers can be launched from outside of Webots.</li>
+=======
 <li>Added the following PROTO objects: 'Radiator', 'WallPlug', 'FireExtinguisher', 'CapScrew', 'ElectricalPlug', 'EyeScrew', 'ScrewHole', 'Washer', 'DoubleFluorescentLamp' and 'OfficeTelephone'.</li>
 <li>Added the following PROTO appearances: 'CorrodedMetal', 'CorrugatedMetal', 'DamascusSteel', 'ElectricConduit', 'FlexibleAluminiumDuct', 'GalvanizedMetal', 'HammeredCopper', 'OsbWood', 'Plaster', 'RustyMetal', 'ScrewThread' and 'WireFence'.</li>
 <li>Added a 'PipeBoundingObject' and a 'TorusBoundingObject' PROTO nodes to simplify the creation of complex bounding objects.</li>
 <li>Added a factory hall environment sample.</li>
+>>>>>>> origin/develop
```
Comment (DavidMansolino) [inline change_logs/ChangeLog.html]:
```suggestion
```

File: change_logs/ChangeLog.html [hunk 2]
```diff
@@ -20,10 +20,14 @@ <h2><a href='https://cyberbotics.com/doc/blog/Webots-2018-b-release'>Webots R201
 <li><b>New Features</b></li>
 <ul>
 <li>Bloom and Ambient Occlusion are now also available as effects for color cameras.</li>
+<<<<<<< HEAD
+<li>Extern robot controllers can be launched from outside of Webots.</li>
+=======
```
Comment (DavidMansolino) [inline change_logs/ChangeLog.html]:
```suggestion
```

File: change_logs/ChangeLog.html [hunk 3]
```diff
@@ -20,10 +20,14 @@ <h2><a href='https://cyberbotics.com/doc/blog/Webots-2018-b-release'>Webots R201
 <li><b>New Features</b></li>
 <ul>
 <li>Bloom and Ambient Occlusion are now also available as effects for color cameras.</li>
+<<<<<<< HEAD
```
Comment (DavidMansolino) [inline change_logs/ChangeLog.html]:
```suggestion
```

File: docs/guide/debugging-c-cpp-controllers.md [hunk 4]
```diff
@@ -34,6 +34,8 @@ The following example assumes that there is a problem with the "soccer\_supervis
 
 ### Using the GNU Debugger with a Controller
 
+On Windows you will need to install MSYS2 and the `mingw-w64-x86_64-gdb` package as indicated in the [optional dependencies](https://github.com/omichel/webots/wiki/Windows-Optional-Dependencies) of the [Windows installation instructions](https://github.com/omichel/webots/wiki/Windows-installation).
```
Comment (DavidMansolino) [inline docs/guide/debugging-c-cpp-controllers.md]: That's probably not the only option (and probably not the simplest, lighter for the user), I would rephrase it as 'On Windows, you can install GDB with MSYS2 for example...'.

File: docs/guide/running-extern-robot-controllers.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+If that simulation has more than one extern controller, you may also set the `WEBOTS_ROBOT_NAME` environment variable to specify the robot to which your controller should connect.
+
+> **Note**: the environment variables can be set inside the controller program, before calling the `wb_robot_init()` function.
+
+## Example Usage
+
+1. Open for example the "WEBOTS\_HOME/projects/robots/softbank/nao/worlds/nao_demo.wbt" world file.
+2. If the simulation was running, stop it and revert it.
+3. Then, open the Nao node in the scene tree and change its controller field from `nao_demo` to `<extern>`.
+4. Save the simulation, restart it and run it.
```
Comment (DavidMansolino) [inline docs/guide/running-extern-robot-controllers.md]: Is this really required?

File: docs/guide/running-extern-robot-controllers.md [hunk 6]
```diff
@@ -0,0 +1,54 @@
+# Running Extern Robot Controllers
+
+This chapter describes extern robot controllers and how to use them.
+
+## Introduction
+
+Normally, Webots launches automatically the robot controller specified in the `controller` field of each [Robot](../reference/robot.md) node.
+However, if this field is set to `<extern>`, no controller is launched and the robot will behave like if its `controller` field was an empty string, that is, the robot will not be controlled.
```
Comment (DavidMansolino) [inline docs/guide/running-extern-robot-controllers.md]: Do we really want the simulation to run while 'waiting' for the controller?
This doesn't seem really reproducible (which is a key element, especially when debugging).

File: docs/guide/running-extern-robot-controllers.md (hunk 5 above)
Comment (omichel) [reply docs/guide/running-extern-robot-controllers.md]: No, but it is convenient. If you don't save, you will have to repeat the operation (setting the controller to `<extern>`) several times.

File: docs/guide/running-extern-robot-controllers.md (hunk 6 above)
Comment (omichel) [reply docs/guide/running-extern-robot-controllers.md]: Yes, I believe this is the best way to go, because it opens all the possibilities to the user:
1. If you want to have a reproducible setup, you simply have to start the simulation in pause mode, connect the controller and run it.
2. If you want to have a world into which controllers can connect any time, while the simulation is running, this is also possible.

**Comment by omichel** [comment]
test me

**Comment by omichel** [comment]
test me

**Comment by omichel** [comment]
Thank you for the tests and review. I believe I fixed all the problems you reported.
👍

**Comment by DavidMansolino** [comment]
:eyes:

**Comment by DavidMansolino** [comment]
Unfortunately, the assertion failure is still present with the exact same stack and procedure:
  - 1. e-puck.wbt world and set the controller to `<extern>`.
  - 2. Start the simulation.
  - 3. form the terminal launch the `e-puck`controller.
```
Thread 1 "webots-bin" received signal SIGABRT, Aborted.
0x00007ffff2859428 in __GI_raise (sig=sig@entry=6)
    at ../sysdeps/unix/sysv/linux/raise.c:54
54	../sysdeps/unix/sysv/linux/raise.c: Aucun fichier ou dossier de ce type.
(gdb) where
#0  0x00007ffff2859428 in __GI_raise (sig=sig@entry=6)
    at ../sysdeps/unix/sysv/linux/raise.c:54
#1  0x00007ffff285b02a in __GI_abort () at abort.c:89
#2  0x00007ffff2851bd7 in __assert_fail_base (fmt=<optimized out>, 
    assertion=assertion@entry=0xc05d53 "mFirstStep", 
    file=file@entry=0xc05d17 "control/WbControlledWorld.cpp", 
    line=line@entry=153, 
    function=function@entry=0xc05ee0 <WbControlledWorld::startControllerFromSocket(WbRobot*, QLocalSocket*)::__PRETTY_FUNCTION__> "void WbControlledWorld::startControllerFromSocket(WbRobot*, QLocalSocket*)") at assert.c:92
#3  0x00007ffff2851c82 in __GI___assert_fail (assertion=0xc05d53 "mFirstStep", 
    file=0xc05d17 "control/WbControlledWorld.cpp", line=153, 
    function=0xc05ee0 <WbControlledWorld::startControllerFromSocket(WbRobot*, QLocalSocket*)::__PRETTY_FUNCTION__> "void WbControlledWorld::startControllerFromSocket(WbRobot*, QLocalSocket*)") at assert.c:101
#4  0x0000000000801484 in WbControlledWorld::startControllerFromSocket (
    this=0x6e99320, robot=0x724b0b0, socket=0x28a77ba0)
    at control/WbControlledWorld.cpp:153
#5  0x0000000000801b93 in WbControlledWorld::addControllerConnection (
    this=0x6e99320) at control/WbControlledWorld.cpp:220
#6  0x0000000000805b56 in QtPrivate::FunctorCall<QtPrivate::IndexesList<>, QtPrivate::List<>, void, void (WbControlledWorld::*)()>::call(void (WbControlledWorld::*)(), WbControlledWorld*, void**) (
```

I don't understand why you don't have it on Windows as it seems quite obvious that this assertion fails:
https://github.com/omichel/webots/blob/feature-start-controllers-outside-of-webots/src/webots/control/WbControlledWorld.cpp#L153

File: docs/guide/running-extern-robot-controllers.md (hunk 6 above)
Comment (DavidMansolino) [reply docs/guide/running-extern-robot-controllers.md]: Thank you for the explanation, that's fine for me. But in that case in the debugging documentation, we should say to the user that it is recommended to pause the simulation (because now we explicitly say the contrary).

**Comment by omichel** [comment]
test me

**Comment by omichel** [comment]
test me

**Comment by omichel** [comment]
👍

**Comment by DavidMansolino** [comment]
:eyes:

File: docs/guide/debugging-c-cpp-controllers.md [hunk 7]
```diff
@@ -50,29 +52,31 @@ $ make
...
 Note that, the *-g* flag should now appear in the compilation line.
-Once you have recompiled the controller, hit the `Pause` and `Reload` buttons.
-This pauses the simulation and reloads the freshly compiled versions of the controller.
-Now find the process ID (PID) of the "soccer\_supervisor" process, using `ps -e` (Linux) or `ps -x` (macOS), or using the *Task Manager* (Windows).
-The PID is in the left-most column of output of `ps` as shown above.
-Then open a terminal and start the debugger by typing:
+Once you have recompiled the controller, you will need to change controller of the [Robot](../reference/robot.md) node to be [extern](running-extern-robot-controllers.md).
+This can be done from the scene tree:
+Hit the `Pause` and `Reset` buttons, set the `controller` field of the Robot node to `<extern>` and save the world file.
+Press the `Run` button in Webots.
```
Comment (DavidMansolino) [inline docs/guide/debugging-c-cpp-controllers.md]:
```suggestion
```

File: docs/guide/running-extern-robot-controllers.md [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+1. Open for example the "WEBOTS\_HOME/projects/robots/softbank/nao/worlds/nao_demo.wbt" world file.
+2. If the simulation was running, stop it and revert it.
+3. Then, open the Nao node in the scene tree and change its controller field from `nao_demo` to `<extern>`.
+4. Save the simulation, restart it and run it.
+5. Open the "WEBOTS\_HOME/projects/robots/softbank/nao/controllers/nao_demo" folder from a terminal.
+6. Add the path to the Webots `libController` in the OS specific library path environment variable:
+On Windows/MSYS2, type: `export PATH=${PATH}:/C/Program\ Files/Webots/msys64/mingw64/bin`.
+On Linux, type `export LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:${WEBOTS_HOME}/lib`.
+On macOS, type `export DYLD_LIBRARY_PATH=${DYLD_LIBRARY_PATH}:/Applications/Webots/lib`.
+You may need to add extra paths to these variables if your controller relies on other shared libraries (for example if it has specific robot window).
```
Comment (DavidMansolino) [inline docs/guide/running-extern-robot-controllers.md]: I believe this deserves its own chapter:
  - `PATH` is not enough, for example in python we should update `PYTHONPATH` to point to `WEBOTS_HOME/lib/python27`
  - For Windows, we should probably document the 'standard' way of adding an environmental variable (probably even the GUI way) and should not bother the user with MSYS2.
  - the example of the robot window doesn't seem to be relevant.

File: docs/guide/running-extern-robot-controllers.md [hunk 9]
```diff
@@ -16,6 +16,33 @@ For example, it may run it within a debugging environment, like *gdb*, a command
 Also, the standard output and error streams (`stdout` and `stderr`) remain under the user control and are not sent to the Webots console.
 It is even possible to read the standard input stream (`stdin`) like with any standard program.
 
+## Environment Variables
+
+In order to be able to run an extern Webots controller, a number of environment variables should be set et extended.
```
Comment (DavidMansolino) [inline docs/guide/running-extern-robot-controllers.md]: It seems there is a type here: 
> be set et extended
